crates/ide_assists/src/handlers/convert_while_to_loop.rs

```diff
@@ -0,0 +1,180 @@
+use std::iter::once;
+
+use syntax::{
+    ast::{
+        self,
+        edit::{AstNodeEdit, IndentLevel},
+        make, LoopBodyOwner,
+    },
+    AstNode,
+};
+
+use crate::{
+    assist_context::{AssistContext, Assists},
+    utils::invert_boolean_expression,
+    AssistId, AssistKind,
+};
+
+// Assist: convert_while_to_loop
+//
+// Replace a while with a loop
+//
+// ```
+// fn main() {
+//     $0while cond {
+//         foo();
+//     }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     loop {
+//         if !cond {
+//             break;
+//         }
+//         foo();
+//     }
+// }
+// ```
+pub(crate) fn convert_while_to_loop(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let while_expr: ast::WhileExpr = ctx.find_node_at_offset()?;
+    let cond = while_expr.condition()?;
+
+    // Don't handle while let
+    if let Some(_) = cond.pat() {
+        return None;
+    };
+
+    let cond_expr = cond.expr()?;
+
+    let target = while_expr.syntax().text_range();
+    acc.add(
+        AssistId("convert_while_to_loop", AssistKind::RefactorRewrite),
+        "Convert while to loop",
+        target,
+        |edit| {
+            let while_expr = edit.make_mut(while_expr);
+            let while_indent_level = IndentLevel::from_node(while_expr.syntax());
+            let while_body = while_expr.loop_body();
+
+            let replacement = {
+                let if_expr = {
+                    let cond = invert_boolean_expression(cond_expr);
+                    let then_branch = make::block_expr(
+                        once(make::expr_stmt(make::expr_break(None)).into()),
+                        None,
+                    );
+
+                    make::expr_if(make::condition(cond, None), then_branch, None)
+                };
+
+                let if_expr = if_expr.indent(while_indent_level);
+
+                let block_expr = if let Some(while_body) = while_body {
+                    let stmts =
+                        once(make::expr_stmt(if_expr).into()).chain(while_body.statements());
+                    make::block_expr(stmts, while_body.tail_expr())
+                } else {
+                    make::block_expr(once(make::expr_stmt(if_expr).into()), None)
```

Comment:
I ran the tests with coverage and it turns out that this line doesn't run when we convert code like
`while$0 cond() {}`.
In other words: in what case `while_expr.loop_body();` is None?

Reply:
The loop_body is none if the body is missing which can happen on parsing recovery I imagine, in that case the assist just shouldn't be applicable.